### DISCUSSION TITLE
What are jj/keep items?

### DISCUSSION THREAD

**Discussion by toomyem** [jj-vcs/jj#4848] [General]
Hi.

I'm using Git Extensions for viewing git repositories. I noticed that jj creates lots of entries named `jj/keep/...` and I understand that those are temporary commits of working space. But I'm wondering, from the technical point of view, what are those exactly? Is it some kind of git tag? And second question - is there a way to clean them, so they don't clutter the view? 

![obraz](https://github.com/user-attachments/assets/75596b5d-4922-4d0a-a2ab-7cc070062c9a)

**Comment by thoughtpolice** [comment]
The git object store is a big, directed graph of references between objects. Occasionally, you need to run garbage collection to remove old objects, when they have nothing pointing to them. To start the process, you begin by sweeping "from the roots", which are known objects you can always start from. You can find the object store under `.git/objects`, which is just a big ol' directory of files.

Git treats visible branches and references as the roots when doing garbage collection; however, this flies in the face of jj, which primarily revolves around unnamed branches and automatic "changeset evolution", which will discard old commits and allow you to return to them. So how do we keep things around?

In reality, Git actually looks at the file called `.git/info/refs` in order to find all the roots. This file is just a simple key-value list, pointing from an object ID to another file. And every file inside of it just contains a single line containing the commit ID of an object in the object store. That object is then a root and won't be GC'd:

```
austin@GANON:~/src/qq$ cat .git/info/refs | grep main
5cca3997c8eb40db6f875388ba13e36e636f0a43        refs/heads/main
5cca3997c8eb40db6f875388ba13e36e636f0a43        refs/remotes/origin/main

austin@GANON:~/src/qq$ cat .git/refs/remotes/origin/main
d30125e90b7be2370645445d42f1e1607ba100f0

austin@GANON:~/src/qq$ cat .git/refs/jj/keep/82ab257550d8117ecec10284cbd1ee22225ad1b4
82ab257550d8117ecec10284cbd1ee22225ad1b4
```

So now things should be pretty clear: jj stores Git references in its "keep" directory, to keep pointers to jj-managed commits that exist in the operation log. That means when your editor decides to helpfully run a background GC or reprune (or Git does it automatically on its own when invoked), your jj repository won't be trashed.

**Comment by thoughtpolice** [comment]
Regarding Git extensions: I'm not quite sure how to make it ignore the jj/keep namespace. Maybe there's a way to filter references shown in the UX? I think most Git GUIs probably only show `refs/heads/` and maybe `refs/remotes` by default, but if you include *everything* then it will include `refs/jj/keep/*`. So maybe there's a way to exclude it? It should not ever be necessary for your day-to-day work, so you can just do a blanket global exclusion.

**Reply by shanesveller** [reply]
For Magit, I can set `git config --local log.excludeDecoration 'refs/jj/keep/*'` to omit them from display without disrupting the actual data storage. I have no idea if this is portable to other tools, but 🤞🏻

**Comment by toomyem** [comment]
Thank you all for your replays. Now I understand what is going on 😄 
For Git Extensions, I can set a filter to "--exclude=refs/jj/keep/*" to hide what is not necessary for me.

Cheers.
